System/Win32/MinTTY.hsc

```diff
@@ -0,0 +1,207 @@
+    len :: DWORD <- peek buf
+    let len' = fromIntegral len `div` sizeOfWchar
+    peekCWStringLen (buf `plusPtr` sizeOfDWORD, min len' mAX_PATH)
+
+getFileInformationByHandleEx
+  :: HANDLE -> CInt -> Ptr a -> DWORD -> IO ()
+getFileInformationByHandleEx h cls buf bufSize = do
+  lib <- getModuleHandle (Just "kernel32.dll")
+  ptr <- getProcAddress lib "GetFileInformationByHandleEx"
+  let c_GetFileInformationByHandleEx =
+        mk_GetFileInformationByHandleEx (castPtrToFunPtr ptr)
+  failIfFalse_ "getFileInformationByHandleEx"
+    (c_GetFileInformationByHandleEx h cls buf bufSize)
+
+ntQueryObjectNameInformation :: HANDLE -> IO String
+ntQueryObjectNameInformation h = do
+  let sizeOfUSHORT = sizeOf (undefined :: USHORT)
+  let sizeOfWchar  = sizeOf (undefined :: CWchar)
+  let bufSize = 8 * sizeOfUSHORT + mAX_PATH * sizeOfWchar
+  allocaBytes bufSize $ \buf ->
+    alloca $ \p_len -> do
+      ntQueryObject h oBJECT_NAME_INFORMATION buf (fromIntegral bufSize) p_len
+      len :: USHORT <- peek buf
+      let len' = fromIntegral len `div` sizeOfWchar
+      peekCWStringLen (buf `plusPtr` (8 * sizeOfUSHORT), min len' mAX_PATH)
+
+ntQueryObject :: HANDLE -> CInt -> Ptr a -> ULONG -> Ptr ULONG -> IO ()
+ntQueryObject h cls buf bufSize p_len = do
+  lib <- getModuleHandle (Just "ntdll.dll")
+  ptr <- getProcAddress lib "NtQueryObject"
+  let c_NtQueryObject = mk_NtQueryObject (castPtrToFunPtr ptr)
+  void $ failIfNeg "NtQueryObject" $ c_NtQueryObject h cls buf bufSize p_len
+
+isVistaOrLater :: IO Bool
+isVistaOrLater = do
+  lib <- getModuleHandle (Just "ntdll.dll")
+  ptr <- getProcAddress lib "RtlGetVersion"
+  let c_RtlGetVersion = mk_RtlGetVersion (castPtrToFunPtr ptr)
+  osVersionInfo <- alloca $ \p -> do
+    _ <- failIfNeg "RtlGetVersion" $ c_RtlGetVersion p
+    peek p
+  return $ dwMajorVersion osVersionInfo >= 6
+
+fILE_NAME_INFO :: CInt
+fILE_NAME_INFO = 2
+
+mAX_PATH :: Num a => a
+mAX_PATH = #const MAX_PATH
+
+oBJECT_NAME_INFORMATION :: CInt
+oBJECT_NAME_INFORMATION = 1
+
+type F_GetFileInformationByHandleEx a =
+  HANDLE -> CInt -> Ptr a -> DWORD -> IO BOOL
+
+foreign import WINDOWS_CCONV "dynamic"
+  mk_GetFileInformationByHandleEx
+  :: FunPtr (F_GetFileInformationByHandleEx a)
+  -> F_GetFileInformationByHandleEx a
+
+type F_NtQueryObject a =
+  HANDLE -> CInt -> Ptr a -> ULONG -> Ptr ULONG -> IO NTSTATUS
+
+foreign import WINDOWS_CCONV "dynamic"
+  mk_NtQueryObject :: FunPtr (F_NtQueryObject a) -> F_NtQueryObject a
+
+type F_RtlGetVersion = Ptr RTL_OSVERSIONINFOW -> IO NTSTATUS
+
+foreign import WINDOWS_CCONV "dynamic"
+  mk_RtlGetVersion :: FunPtr F_RtlGetVersion -> F_RtlGetVersion
+
+type ULONG    = #type ULONG
+type NTSTATUS = #type NTSTATUS
+
+data RTL_OSVERSIONINFOW = RTL_OSVERSIONINFOW
+  { dwOSVersionInfoSize :: ULONG
+  , dwMajorVersion      :: ULONG
+  , dwMinorVersion      :: ULONG
+  , dwBuildNumber       :: ULONG
+  , dwPlatformId        :: ULONG
+  , szCSDVersion        :: CWString
+  } deriving Show
```

Comment:
The ability to query your Windows version might be handy as its own feature. Thoughts?